qc_schema/dev/dev_schema.py

```diff
@@ -37,9 +37,20 @@
                 },
                 "basis": {
                     "type": "string"
+                },
+                "basis_spec": {
+                    "$ref": "#/definitions/basis_spec"
                 }
             },
-            "required": ["method", "basis"],
+            "required": [ "method" ],
+            "oneOf": [
+                {
+                    "required": ["basis"]
+                },
+                {
+                    "required": ["basis_spec"]
+                }
+            ],
```

Comment:
Is this sufficient to ensure that a O-H-H-Zr molecule can't be validated with just a `{'basis': 1, 'basis_spec': 4}` (that's meant to provide basis spec for all hydrogens and atom # 4 (Zr))? That is, would this catch that atom # 1 is empty? Or is that the responsibility of some other part of the spec?

Reply:
This gets into a rather fun debate that a schema really only determines the structure of information, it cannot really verify that information to any detail. JSON-Schema attempts to go slightly deeper into the verification, but there is no way to encode this particular logic as far as I know. The benefit of using pure JSON-Schema is that the validator is available in most languages already.

Reply:
I think that has to be handled in code. I suppose it is also possible to have an atom or center without basis functions (dummy center?).

Reply:
Ah, structure, not verification. That's useful to have pounded into my head. So JSON-Schema is rather doing the type-check and necessary-args check of the validation step. Real code does validation and fill-in-the-holes. Makes sense, as I wouldn't have expected JSON-Schema to handle the detailed Molecule validation either.

Reply:
+1

In cclib we've encountered this issue multiple times: do we try to catch nonsensical numbers/results from parsed logfiles? The answer is always emphatically **no**. It's simply outside the purview of parsing, and I think the situation is similar with the schema. It should do what it's meant to as well as possible, namely transporting data, but not much more.